functorch/csrc/BatchRulesBinaryOps.cpp

```diff
@@ -218,6 +297,10 @@ TORCH_LIBRARY_IMPL(aten, FT_BATCHED_KEY, m) {
   UNARY_POINTWISE(clamp_max);
   POINTWISE_BOXED(clamp_max_);
 
+  VARIADIC_BDIMS_BOXED(_euclidean_dist);
+  BINARY_POINTWISE_CONTIG(_cdist_forward);
```

Comment:
Why do we need to make the inputs for this contiguous?

If the backward pass requires saved inputs to be contiguous, then we should handle that in the backward batch rule (.contiguous() is generally an expensive operation and if a user never calls .backward() then we don't want them to pay for it)

Reply:
Original cdist op is working such that forward pass does not fail on non-contig input but backward pass fails as requires contiguous input:
```python
x = torch.rand(4, 32, 5)[:, 0, :]
y = torch.rand(6, 32, 5)[:, 0, :]
assert not x.is_contiguous()
assert not y.is_contiguous()

x.requires_grad_()
_ = y.requires_grad_()

ret = torch.cdist(x, y)
ret = ret.sum()
ret.backward()

> RuntimeErrorTraceback (most recent call last)
<ipython-input-44-c6bee6e0f2fc> in <module>
      9 ret = torch.cdist(x, y)
     10 ret = ret.sum()
---> 11 ret.backward()

/opt/conda/lib/python3.8/site-packages/torch/_tensor.py in backward(self, gradient, retain_graph, create_graph, inputs)
    305                 create_graph=create_graph,
    306                 inputs=inputs)
--> 307         torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
    308 
    309     def register_hook(self, hook):

/opt/conda/lib/python3.8/site-packages/torch/autograd/__init__.py in backward(tensors, grad_tensors, retain_graph, create_graph, grad_variables, inputs)
    152         retain_graph = create_graph
    153 
--> 154     Variable._execution_engine.run_backward(
    155         tensors, grad_tensors_, retain_graph, create_graph, inputs,
    156         allow_unreachable=True, accumulate_grad=True)  # allow_unreachable flag

RuntimeError: _cdist_backward requires X1 to be contiguous
```

Maybe, original backward pass could simply call `.contiguous()` on non-contig input inside backward pass instead of failing in pytorch.

From functorch side, the ops like vjpvmap do not call backward batch rule but only forward batch rule and original backward pass and thus failing with the same as above message.

Reply:
Oh yeah, this is a PyTorch bug

Reply:
https://github.com/pytorch/pytorch/issues/69997 --> We should have cdist_backward call .contiguous in PyTorch. In general, if an operation is supported in the forward pass by PyTorch, then the backward pass should also be supported.